### ISSUE TITLE
robots.txt Disallow not respected

### ISSUE THREAD

**Issue by ronjakoi** [Norconex/collector-http#457]
I have a setup like this in my `robots.txt`:

```
User-agent: my_crawler
Crawl-delay: 1

User-agent: * # match all bots
Crawl-delay: 10 # per http://en.wikipedia.org/wiki/Robots.txt#Nonstandard_extensions, sets number of seconds to wait between requests to 5 seconds. may not work
Request-rate: 1/5 # per http://en.wikipedia.org/wiki/Robots.txt#Extended_Standard, maximum rate is one page every 5 seconds. may not work
#--snip lots of directives--
Disallow: /login.action # remove the login page
Disallow: /login.action? # remove the login page derivatives
#--snip some more--
```

Log snippet from my web server:

```
crawler.ip.address - - [15/Jan/2018:23:17:39 +0200] "GET /login.action?os_destination=%2F HTTP/1.1" 200 58895
```

Corresponding snippet from the Collector's log:

```
wiki: 2018-01-15 23:17:39 INFO -          DOCUMENT_FETCHED: https://wiki.mydomain.fi/login.action?os_destination=%2F
wiki: 2018-01-15 23:17:39 INFO -       CREATED_ROBOTS_META: https://wiki.mydomain.fi/login.action?os_destination=%2F
wiki: 2018-01-15 23:17:39 INFO - REJECTED_ROBOTS_META_NOINDEX: https://wiki.mydomain.fi/login.action?os_destination=%2F
```

My config:

```xml
<httpcollector id="mycrawler">
    #set($http      = "com.norconex.collector.http")
    #set($robotsTxt         = "${http}.robot.impl.StandardRobotsTxtProvider")
    <crawlerDefaults>
        <robotsTxt ignore="false" class="$robotsTxt"/>
        <!-- ... -->
    </crawlerDefaults>
    
    <!-- ... -->
</httpcollector>
```

I am using HTTP Collector 2.7.1.

**Comment by essiembre** [comment]
The ``REJECTED...`` log entry is about the "meta" (or "tag") robot rules, not your ``robots.txt``.  That means in the HTTP Response Header or within the HTML page itself, there are instructions not to index the page.  See [StandardRobotsMetaProvider](https://www.norconex.com/collectors/collector-http/latest/apidocs/com/norconex/collector/http/robot/impl/StandardRobotsMetaProvider.html) for details.  

In short, if you want to disable support for robot meta/tag support, you can do so with:
```xml
<robotsMeta ignore="true"/>
```

Please confirm.

**Comment by ronjakoi** [comment]
I am aware of this. I'm afraid you have misunderstood. My issue is that the `DOCUMENT_FETCHED` event should have never occurred, as my `robots.txt` contains the directive `Disallow: /login.action`. The fact that the HTML response or meta tags in the document also disallow this URI is coincidental.

**Comment by essiembre** [comment]
Haaa.. got it now.  Do you have a URL you can share to reproduce?

**Comment by ronjakoi** [comment]
I emailed you some.

**Comment by essiembre** [comment]
It works for me when I try it.   I use the ``XMLFileCommiter``.  Can you try with the following config and see if it gets rejected as it should?

```xml
<?xml version="1.0" encoding="UTF-8"?>
<httpcollector id="HTTP Collector">

  #set($workdir = "./workdir/issue457")

  <progressDir>$workdir</progressDir>
  <logsDir>$workdir</logsDir>

  <crawlers>
    <crawler id="HTTP Crawler">
    
      <numThreads>1</numThreads>
      <orphansStrategy>IGNORE</orphansStrategy>
      <delay default="500" />
      <robotsTxt ignore="false"/>
      <workDir>$workdir</workDir>
      <maxDepth>2</maxDepth>
      <maxDocuments>2</maxDocuments>
      <canonicalLinkDetector ignore="true" />
      <sitemapResolverFactory ignore="false"/>

      <startURLs stayOnDomain="true" stayOnPort="true" stayOnProtocol="true">
        <url>[YOUR_FAULTY_URL]</url> 
      </startURLs>

      <committer class="com.norconex.committer.core.impl.XMLFileCommitter">
        <directory>${workdir}/committer-xml</directory>
        <pretty>true</pretty>
        <docsPerFile>100</docsPerFile>
      </committer>
      
    </crawler>
  </crawlers>
</httpcollector>
```

**Comment by ronjakoi** [comment]
Confirmed. With this config my problem URLs get rejected with a `REJECTED_ROBOTS_TXT` as expected.

In my environment I tried this:

```
[ronjakoi@crawler logs]$ pwd
/opt/norconex/mydomain.fi/daily/work/logs
[ronjakoi@crawler logs]$ grep -r 'REJECTED_ROBOTS_TXT: https://wiki'
```

and got zero hits. Quite puzzling.

**Comment by essiembre** [comment]
It points to something odd in your config.  If you have a way to share it, that would be nice.  Otherwise, proceed by elimination and add to that working config your own config, one item at a time until you reproduce the issue.

One thing maybe.... do you have a mix of settings defined under ``crawlerDefaults`` and ``crawler``?  Maybe some of the default settings are overwritten when they should not, or the opposite: you have defaults you should overwrite.

**Comment by ronjakoi** [comment]
I narrowed it down. My `robots.txt` contains a section that is an exact match for my crawler's `userAgent`:
```
User-Agent: my_crawler
Crawl-delay: 1
```

This section sets no `Disallow` directives. However there is another section targeting all user-agents and containing many `Disallow` directives:

```
User-Agent: *
Crawl-delay: 10
Disallow: [REDACTED]
-snip-
```

If I unset `<userAgent>my_crawler</userAgent>` or change it to string that doesn't match the first section, I get a `REJECTED_ROBOTS_TXT` event as expected.

**Comment by essiembre** [comment]
Good catch.  Looks like it works as expected then. Can we close this issue?

**Comment by ronjakoi** [comment]
No. Your implementation is wrong. The problem is in [`StandardRobotsTxtProvider`](https://github.com/Norconex/collector-http/blob/e7f7e239588adc6223c470d9219cb7bcaa25f475/norconex-collector-http/src/main/java/com/norconex/collector/http/robot/impl/StandardRobotsTxtProvider.java#L132).

Compare to http://www.robotstxt.org/norobots-rfc.txt, lines 427–486:

```
4. Examples

   This section contains an example of how a /robots.txt may be used.

   A fictional site may have the following URLs:

     http://www.fict.org/
     http://www.fict.org/index.html
     http://www.fict.org/robots.txt
     http://www.fict.org/server.html
     http://www.fict.org/services/fast.html
     http://www.fict.org/services/slow.html
     http://www.fict.org/orgo.gif
     http://www.fict.org/org/about.html
     http://www.fict.org/org/plans.html
     http://www.fict.org/%7Ejim/jim.html
     http://www.fict.org/%7Emak/mak.html

   The site may in the /robots.txt have specific rules for robots that
   send a HTTP User-agent "UnhipBot/0.1", "WebCrawler/3.0", and


Koster                draft-koster-robots-00.txt                [Page 8]

INTERNET DRAFT        A Method for Robots Control       December 4, 1996

   "Excite/1.0", and a set of default rules:

      # /robots.txt for http://www.fict.org/
      # comments to webmaster@fict.org

      User-agent: unhipbot
      Disallow: /

      User-agent: webcrawler
      User-agent: excite
      Disallow: 

      User-agent: *
      Disallow: /org/plans.html
      Allow: /org/
      Allow: /serv
      Allow: /~mak
     Disallow: /

   The following matrix shows which robots are allowed to access URLs:

                                               unhipbot webcrawler other
                                                        & excite
     http://www.fict.org/                         No       Yes       No
     http://www.fict.org/index.html               No       Yes       No
     http://www.fict.org/robots.txt               Yes      Yes       Yes
     http://www.fict.org/server.html              No       Yes       Yes
     http://www.fict.org/services/fast.html       No       Yes       Yes
     http://www.fict.org/services/slow.html       No       Yes       Yes
     http://www.fict.org/orgo.gif                 No       Yes       No
     http://www.fict.org/org/about.html           No       Yes       Yes
     http://www.fict.org/org/plans.html           No       Yes       No
     http://www.fict.org/%7Ejim/jim.html          No       Yes       No
     http://www.fict.org/%7Emak/mak.html          No       Yes       Yes
```

Norconex HTTP Collector explicitly stops parsing after a block with an exact-match `User-agent`. This is not correct behavior, it should also parse all `User-agent: *` blocks.

**Comment by essiembre** [comment]
Where do you get it is wrong? If there is a match for your user-agent, that is the record block to be used.  ONLY if there is no match for it that ``User-agent: *`` gets used.  This is described in the same specs you mention:

> The robot must obey the first record in /robots.txt that contains a User-
> Agent line whose value contains the name token of the robot as a 
> substring. The name comparisons are case-insensitive. If no such
> record exists, it should obey the first record with a User-agent
> line with a "*" value, if present. If no record satisfied either
> condition, or no records are present at all, access is unlimited.

This is also what the examples you copied are illustrating.  If you want ``my_crawler`` to disallow some paths, you have to include those under ``my_crawler``.

This is also reinforced by Google (https://developers.google.com/search/reference/robots_txt):

> Only one group of group-member records is valid for a particular crawler. The crawler must determine the correct group of records by finding the group with the most specific user-agent that still matches. All other groups of records are ignored by the crawler.

Unless you can find evidence of the contrary in the specifications, I am afraid you will have to modify your robots.txt or change your user-agent (or apply filtering rules yourself in the Collector config).

Please confirm.

**Comment by ronjakoi** [comment]
Confirmed, you are correct. I have read the documentation incorrectly. Apologies for my previous, unnecessarily blunt message, as well!
